### ISSUE TITLE
Fix S2955 FP: following rule suggestion causes compiler error

### ISSUE THREAD

**Issue by paulhickman-ec** [SonarSource/sonar-dotnet#9493]
### Description

The following code causes S2955. However comparing to default(T) instead of null as the rule suggests gives a compiler error. Even if it did compile, if T was int then default(T) would be zero which gives different behavior than comparing T? to null.

### Repro steps

````c#
public bool Foo<T>(Func<T?> bar) where T : notnull
{
    T? t = bar();
    return t == null;
}
````

### Expected behavior

Rule does not fire

### Actual behavior

Rule does fire

### Known workarounds

Ignore the rule

### Related information

* SonarAnalyzer.CSharp 9.28.0.94264
* .Net 8

**Comment by jilles-sg** [comment]
In this particular case, the rule is correct, but perhaps not for the right reason. The `notnull` constraint allows both value and reference types and therefore `T?` is the same as `T` at run time. This can be seen by trying to compile
```csharp
bool result = Foo<int>(() => (int?)null);
```
which fails with `CS0266: Cannot implicitly convert type 'int?' to 'int'.`.

On the other hand, if we let the compiler infer, it will infer `int?` and not `int`, resulting in warning CS8714 that `int?` does not match the constraint. It will probably do what was expected, though.

Trying to make this work in the general case is not so easy. If the non-nullable type is not needed, making `T` the nullable type would work (but S2955 will probably have to be suppressed). Otherwise, making the constraint `where T: struct` or `where T: class` would work (excluding either reference or value types).

**Comment by jeremysalwen** [comment]
FWIW here is a more practical usage encountering the same false positive:

```
public static class IEnumerableExtensions
{
    public static IEnumerable<TValue>
        SelectNonNull<T, TValue>(
            this IEnumerable<T> source,
            Func<T, TValue?> valueSelector)
        where TValue : notnull =>
        source
            .Select(valueSelector)
            .Where(e => e != null)
            .Select(e => e!);
}
```

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1575](https://sonarsource.atlassian.net/browse/NET-1575)

[NET-1575]: https://sonarsource.atlassian.net/browse/NET-1575?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
